ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,34 @@
+# Choosing a Versioning Schemes
+
+This document records our discussions around choosing how to version our packages.
+
+## Desiderata
+
+- *Simplicity and Familiarity*. We'd like the versioning scheme to be simple to explain, and ideally already well-established.
+
+- *Ease of execution*. We'd like the process of cutting a release to merely involve following a very simple checklist. We'd like it to require as few inputs, discussions, decisions, etc.
+
+- *Distinguished development versions*. We'd like to distinguish between the two possible semantics of version number.
+
+    - The version of a released thing identifies some immutable thing, which is always somehow more refined than any release that has a lesser version number.
+
+    - A development version refers to some _mutable_ thing that is improving during the time between two releases, eg the version on master. IE there will usually be multiple different commits that all have the same development version number.
+
+## Proposal Simplest
+
+To cut a release, merely create branch release/XXX pointing at the desired commit on master and also immediately create a subsequent master commit that advances all the appropriate versions.
+
+## Proposal Parity
+
+Minor versions will be odd for packages on master and even for releases (like the GHC Team's scheme).
+To cut a release, create branch release/XXX pointing at the desired master commit and then add a commit to release/XXX that bumps all versions to the next even number, and also immediately create a subsequent master commit that advances all the appropriate versions to the next odd number.
+
+## Proposal NonZero
+
+Master always has degenerate versions on it: everything is version 0.
+To cut a release, create branch release/XXX pointing at the desired master commit and then add a commit to release/XXX that advances all appropriate versions COMPARED TO their value in the previous release branch.
+
+## Proposal Dimension
```

Comment:
I currently favor `Proposal Dimension`, FWIW.

Reply:
I favor this one too! It also allows adding patch on release (hotfix).

Reply:
> It also allows adding patch on release (hotfix).

I think whether we are able to create patch releases is mostly orthogonal to the different proposals here -- i.e. Proposal Dimension and Proposal NonZero only differ in what version is written down in the `.cabal` files on `master`, but the versions in CHaP would/could be exactly the same.

Reply:
I also like this one the most, it's quite clean.  It has the nice side effect that all your packages will have the same `major` version if I understand correctly.  This makes it quite nice for downstream dependencies.  Further more if you will have some reasonable constraints between all your packages, then declaration of which version of `consensus` is used might be as simple as constraining only one of `consensus` packages in `cardano-node` repo.

Reply:
I'm also ok with this proposal.

Reply:
However, I like this option better:

>  Whenever a change is done on master (or a feature branch before), consider how it impacts the version number, i.e. bump the corresponding digit for breaking changes or patches when you do break an interface or just patch something. The master version then indicates the version number which would be true if you just release today (e.g. by tagging a commit).